numpy/core/src/multiarray/experimental_public_dtype_api.c

```diff
@@ -180,11 +180,78 @@ PyArrayInitDTypeMeta_FromSpec(
          * It is up to the user to get this right, and slots are sorted
          * exactly like they are stored right now:
          */
-        void **current = (void **)(&(
-                NPY_DT_SLOTS(DType)->discover_descr_from_pyobject));
-        current += slot - 1;
-        *current = pfunc;
+        if (slot <= 8) {
+            // slot > 8 are PyArray_ArrFuncs
+            void **current = (void **)(&(
+                    NPY_DT_SLOTS(DType)->discover_descr_from_pyobject));
+            current += slot - 1;
+            *current = pfunc;
+        }
+        else {
+            switch (slot) {
+                case 0 + (1 << 10):
+                    NPY_DT_SLOTS(DType)->f.getitem = pfunc;
+                    break;
+                case 1 + (1 << 10):
+                    NPY_DT_SLOTS(DType)->f.setitem = pfunc;
+                    break;
+                case 2 + (1 << 10):
+                    NPY_DT_SLOTS(DType)->f.copyswapn = pfunc;
+                    break;
+                case 3 + (1 << 10):
+                    NPY_DT_SLOTS(DType)->f.copyswap = pfunc;
+                    break;
+                case 4 + (1 << 10):
+                    NPY_DT_SLOTS(DType)->f.compare = pfunc;
+                    break;
+                case 5 + (1 << 10):
+                    NPY_DT_SLOTS(DType)->f.argmax = pfunc;
+                    break;
+                case 6 + (1 << 10):
+                    NPY_DT_SLOTS(DType)->f.dotfunc = pfunc;
+                    break;
+                case 7 + (1 << 10):
+                    NPY_DT_SLOTS(DType)->f.scanfunc = pfunc;
+                    break;
+                case 8 + (1 << 10):
+                    NPY_DT_SLOTS(DType)->f.fromstr = pfunc;
+                    break;
+                case 9 + (1 << 10):
+                    NPY_DT_SLOTS(DType)->f.nonzero = pfunc;
+                    break;
+                case 10 + (1 << 10):
+                    NPY_DT_SLOTS(DType)->f.fill = pfunc;
+                    break;
+                case 11 + (1 << 10):
+                    NPY_DT_SLOTS(DType)->f.fillwithscalar = pfunc;
+                    break;
+                case 12 + (1 << 10):
+                    *NPY_DT_SLOTS(DType)->f.sort = pfunc;
+                    break;
+                case 13 + (1 << 10):
+                    *NPY_DT_SLOTS(DType)->f.argsort = pfunc;
```

Comment:
My thought was that these are all pointers and all so that er could just write:
```
((void *)(*NPY_DT_SLOTS(DType)->f) + num) = pfunc;
```
And not repeat all of this.  Of course you would still need the list of undefined cases to error for, so I am OK with it also.

Reply:
I'd be okay with this except

1. The pointer arithmetic gets a bit tricky here, and if you're not perfectly comfortable you'd get lost - I know I did. If this is to be used by people making their own dtypes (and debugging their own dtypes) then it's probably clearer for them if we use cases.
2. [This array of `PyArray_VectorUnaryFunc*`](https://github.com/peytondmurray/numpy/blob/2d0853d4a454b198e91c20eec705e7bc8ae148d5/numpy/core/include/numpy/ndarraytypes.h#L452) takes up `NPY_NTYPES_API_COMPATIBLE*sizeof(void *)` bytes at the beginning of the struct. We can of course do the arithmetic to account for this, but I think the meaning of the code starts to become more obscure the more fiddling we do. We'd have to have a comment explaining about why we're adding on some strange constant to skip over this array because of how the struct is arranged in memory, and it just seems like a bit much.

So for these two reasons I've left a bunch of cases, which is indeed some repeated code. What do you think?